File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()
        
        for i in range(len(nums)-2):
            if nums[i] == nums[i-1] and i > 0: continue
            
            left = i + 1
            right = len(nums) - 1
            
            while left < right:
                sum = nums[i] + nums[left] + nums[right]
                
                if sum > 0: right -= 1
                elif sum < 0 : left += 1
                else:
                    result.append([nums[i], nums[left], nums[right]])
                    
                    while left < right and nums[left] == nums[left+1]:
                        left += 1
                    while left < right and nums[right-1] == nums[right]:
                        right -= 1
                    
                    left += 1
                    right -= 1
        
        return result
                

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());
        int n = nums.size();
        // bad cases
        if (nums[0] > 0 || n < 3) {
            return {};
        }
        
        for (int i = 0; i < n; i++) {
            int left = i + 1, right = n - 1;
            // ignore any duplicates
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            
            while (left < right) {
                int curSum = nums[i] + nums[left] + nums[right];
                if (curSum > 0) right--;
                else if (curSum < 0) left++;
                else {
                    result.push_back({nums[i], nums[left], nums[right]});
                    while (left < right && nums[left] == nums[left + 1]) left++;  // ignore any duplicates
                    while (left < right && nums[right] == nums[right - 1]) right--;  // ignore any duplicates
                    left++;
                    right--;
                }
            }
        }
        return result;
    }
};